src/spox/_attributes.py

```diff
@@ -184,38 +200,56 @@ def maybe(
     ) -> Optional[AttrIterableT]:
         return cls(tuple(value)) if value is not None else None
 
+    def _to_onnx_deref(self, key: str) -> AttributeProto:
+        return make_attribute(key, self.value, attr_type=self._attribute_proto_type_int)
+
 
 class AttrFloat32s(_AttrIterable[float]):
     _attribute_proto_type_int = AttributeProto.FLOATS
 
     def _to_onnx_deref(self, key: str) -> AttributeProto:
-        # ensure values are all floats
-        values = [float(v) for v in self.value]
-        return make_attribute(key, values, attr_type=AttributeProto.FLOATS)
+        try:
+            transformed = tuple(float(v) for v in self.value)
+        except ValueError as e:
+            raise AttributeTypeError("Attribute values don't seem to be floats.") from e
+        return make_attribute(
+            key, transformed, attr_type=self._attribute_proto_type_int
+        )
 
 
 class AttrInt64s(_AttrIterable[int]):
     _attribute_proto_type_int = AttributeProto.INTS
 
-    def _to_onnx_deref(self, key: str) -> AttributeProto:
-        return make_attribute(key, self.value, attr_type=AttributeProto.INTS)
-
 
 class AttrStrings(_AttrIterable[str]):
     _attribute_proto_type_int = AttributeProto.STRINGS
 
     def _to_onnx_deref(self, key: str) -> AttributeProto:
+        try:
+            transformed = tuple(v.encode() for v in self.value)
+        except AttributeError as e:
+            raise AttributeTypeError(
+                "Attribute values don't seem to be strings."
+            ) from e
         return make_attribute(
-            key, [v.encode() for v in self.value], attr_type=AttributeProto.STRINGS
+            key, transformed, attr_type=self._attribute_proto_type_int
         )
 
 
 class AttrTensors(_AttrIterable[np.ndarray]):
     _attribute_proto_type_int = AttributeProto.TENSORS
 
     def _to_onnx_deref(self, key: str) -> AttributeProto:
-        tensors = [from_array(t) for t in self.value]
-        return make_attribute(key, tensors, attr_type=AttributeProto.TENSORS)
+        try:
+            transformed = tuple(from_array(v) for v in self.value)
+        except AttributeError as e:
+            raise AttributeTypeError(
+                "Attribute values don't seem to be numpy arrays."
```

Comment:
I'm afraid there are plenty of NumPy arrays which will raise an error with `onnx.numpy_helpers.from_array`. Take the following for example:
```python
from_array(np.array(["foo"]))
>>> RuntimeError: Numpy data type not understood yet: dtype('<U3')
```
Come to think of it, we should probably address that issue, too, but my point is that the error message may point the user into the wrong direction.